src/Microsoft.Framework.Caching.SqlServer/SqlServerCache.cs

```diff
@@ -0,0 +1,478 @@
+            // get and cache the ordinal positions of the columns for better performance.
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                var command = new SqlCommand(_sqlQueries.GetTableSchema, connection);
+                await connection.OpenAsync();
+
+                var reader = await command.ExecuteReaderAsync(CommandBehavior.SchemaOnly);
+
+                _cacheItemIdIndex = reader.GetOrdinal(CacheItemId);
+                _cacheItemValueIndex = reader.GetOrdinal(CacheItemValue);
+                _expiresAtTimeUTCIndex = reader.GetOrdinal(ExpiresAtTimeUTC);
+                _slidingExpirationInTicksIndex = reader.GetOrdinal(SlidingExpirationInTicks);
+                _absoluteExpirationIndex = reader.GetOrdinal(AbsoluteExpiration);
+            }
+        }
+
+        public byte[] Get([NotNull] string key)
+        {
+            byte[] cacheItemvalue = null;
+            var utcNow = _systemClock.UtcNow;
+
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                var command = new SqlCommand(_sqlQueries.GetCacheItem, connection);
+                command.Parameters.AddWithValue(CacheItemId, key);
+                command.Parameters.AddWithValue("UtcNow", utcNow);
+
+                connection.Open();
+
+                var reader = command.ExecuteReader(CommandBehavior.SingleRow);
+
+                DateTimeOffset oldExpirationTime;
+                TimeSpan? slidingExpiration = null;
+                DateTimeOffset? absoluteExpiration = null;
+                if (reader.Read())
+                {
+                    oldExpirationTime = reader.GetFieldValue<DateTimeOffset>(_expiresAtTimeUTCIndex);
+                    cacheItemvalue = reader.GetFieldValue<byte[]>(_cacheItemValueIndex);
+
+                    if (!reader.IsDBNull(_slidingExpirationInTicksIndex))
+                    {
+                        slidingExpiration = TimeSpan.FromTicks(
+                            reader.GetFieldValue<long>(_slidingExpirationInTicksIndex));
+                    }
+
+                    if (!reader.IsDBNull(_absoluteExpirationIndex))
+                    {
+                        absoluteExpiration = reader.GetFieldValue<DateTimeOffset>(_absoluteExpirationIndex);
+                    }
+                }
+                else
+                {
+                    return null;
+                }
+
+                // Check if the current item has a sliding expiration and is going to expire within the
+                // next timeout period(ex:20 minutes), then extend the expiration time.
+                if (slidingExpiration.HasValue &&
+                    utcNow >= (oldExpirationTime - slidingExpiration.Value))
+                {
+                    var newExpirationTime = GetExpirationTime(slidingExpiration, absoluteExpiration, utcNow);
```

Comment:
An alternative to the optimization that doubles the expiration time is to send the SELECT and UPDATE/DELETEs necessary to maintain the database updated together in one batch. It doesn't get rid of the overhead of executing more than just the SELECT every time you try to get and item, but at least it gets rid of the extra database round trips.